File: blockchain-ts/src/blockchain.ts
import crypto from 'crypto';
import fetch from 'node-fetch';
import humps from 'humps';
import stringify from 'json-stable-stringify';

type Transaction = {
  sender: string;
  recipient: string;
  amount: number;
}

type Block = {
  index: number;
  timestamp: number;
  transactions: Transaction[];
  proof: number;
  previousHash: string;
}

class Blockchain {
  public chain: Block[] = [];
  private currentTransactions: Transaction[] = [];
  public nodes = new Set<string>();

  public constructor() {
    this.newBlock(100, '1');
  }

  /**
   * Add a new node to the list of nodes
   * @param address Address of node. Eg. 'http://192.168.0.5:5000'
   */
  public registerNode(address: string) {
    try {
      if (address.startsWith('https://'))
        address = `http://${address}`
      const parsedUrl = new URL(address);
      if (parsedUrl.host)
        this.nodes.add(parsedUrl.host);
    } catch (_) {
      throw new Error('Invalid URL');
    }
  }

  /**
   * Determine if a given blockchain is valid
   * @param chain A blockchain
   * @return true if valid, false if not
   */
  public static validChain(chain: Block[]) {
    let lastBlock = chain[0];
    let currentIndex = 1;

    while (currentIndex < chain.length) {
      const block = chain[currentIndex];
      console.log(lastBlock);
      console.log(block);
      console.log('\n--------------\n');

      const lastBlockHash = Blockchain.hash(lastBlock);
      if (block.previousHash != lastBlockHash)
        return false;

      if (!Blockchain.validProof(lastBlock.proof, block.proof, lastBlockHash))
        return false;

      lastBlock = block;
      currentIndex++;
    }

    return true;
  }

  /**
   * This is our consensus algorithm, it resolves conflicts by replacing our chain with the longest one in the network.
   * @return true if our chain was replaced, false if not
   */
  public async resolveConflicts() {
    const neighbors = this.nodes;
    let newChain: Block[] | undefined;

    let maxLength = this.chain.length;

    for (const node of neighbors) {
      try {
        const res = await fetch(`http://${node}/chain`);

        if (res.status === 200) {
          const json: any = await res.json();
          const length: number = json['length'];
          const chain = humps.camelizeKeys(json['chain']) as Block[];

          if (length > maxLength && Blockchain.validChain(chain)) {
            maxLength = length;
            newChain = chain;
          }
        }
      } catch (error) {
        console.log(error);
      }
    }

    if (newChain !== undefined) {
      this.chain = newChain;
      return true;
    }

    return false;
  }

  /**
   * Create a new Block in the Blockchain
   * @param proof The proof given by the Proof of Work algorithm
   * @param previousHash Hash of previous Block
   * @return New Block
   */
  public newBlock(proof: number, previousHash: string) {
    const block = {
      index: this.chain.length + 1,
      timestamp: Date.now() / 1000,
      transactions: this.currentTransactions,
      proof: proof,
      previousHash: previousHash || Blockchain.hash(this.chain[this.chain.length - 1]),
    };

    this.currentTransactions = [];

    this.chain.push(block);
    return block;
  }

  /**
   * Creates a new transaction to go into the next mined Block
   * @param sender Address of the Sender
   * @param recipient Address of the Recipient
   * @param amount Amount
   * @return The index of the Block that will hold this transaction
   */
  public newTransaction(sender: string, recipient: string, amount: number) {
    this.currentTransactions.push({
      sender: sender,
      recipient: recipient,
      amount: amount,
    });

    return this.lastBlock.index + 1;
  }

  /**
   * Creates a SHA-256 hash of a Block
   * @param block Block
   */
  public static hash(block: Block) {
    const blockString = stringify(humps.decamelizeKeys(block));
    const hash = crypto.createHash('sha256');
    return hash.update(blockString).digest('hex');
  }

  public get lastBlock() {
    return this.chain[this.chain.length - 1];
  }

  /**
   * Simple Proof of Work Algorithm:
   * - Find a number p' such that hash(pp') contains leading 4 zeroes
   * - Where p is the previous proof, and p' is the new proof
   * @param lastBlock last Block
   * @return
   */
  public static proofOfWork(lastBlock: Block) {
    const lastProof = lastBlock.proof;
    const lastHash = Blockchain.hash(lastBlock);

    let proof = 0;
    while (!Blockchain.validProof(lastProof, proof, lastHash))
      proof++;

    return proof;
  }

  /**
   * Validates the Proof
   * @param lastProof Previous Proof
   * @param proof Current Proof
   * @param lastHash The hash of the Previous Block
   * @return true if correct, false if not.
   */
  public static validProof(lastProof: number, proof: number, lastHash: string) {
    const guess = `${lastProof}${proof}${lastHash}`;
    const hash = crypto.createHash('sha256');
    const guessHash = hash.update(guess).digest('hex');
    return guessHash.startsWith('0000');
  }
}

export { Transaction, Block, Blockchain }

File: blockchain-go/blockchain/blockchain.go
package blockchain

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type Transaction struct {
	Sender    string `json:"sender"`
	Recipient string `json:"recipient"`
	Amount    int64  `json:"amount"`
}

type Block struct {
	Index        int32          `json:"index"`
	Timestamp    float64        `json:"timestamp"`
	Transactions []*Transaction `json:"transactions"`
	Proof        int64          `json:"proof"`
	PreviousHash string         `json:"previous_hash"`
}

type Blockchain struct {
	currentTransactions []*Transaction
	Chain               []*Block
	Nodes               map[string]bool
}

func New() *Blockchain {
	b := &Blockchain{
		currentTransactions: make([]*Transaction, 0),
		Nodes:               make(map[string]bool),
	}
	b.NewBlock(100, "1")
	return b
}

/*
Add a new node to the list of nodes

 address: Address of node. Eg. 'http://192.168.0.5:5000'
*/
func (b *Blockchain) RegisterNode(address string) error {
	if !strings.HasPrefix(address, "http://") {
		address = "http://" + address
	}
	if parsedUrl, _ := url.Parse(address); parsedUrl.Host != "" {
		b.Nodes[parsedUrl.Host] = true
	} else {
		return errors.New("Invalid URL")
	}
	return nil
}

/*
Determine if a given blockchain is valid

 chain: A blockchain
 return: true if valid, false if not
*/
func ValidChain(chain []*Block) bool {
	lastBlock := chain[0]
	currentIndex := 1

	for currentIndex < len(chain) {
		block := chain[currentIndex]
		lastBlockBytes, _ := json.Marshal(lastBlock)
		blockBytes, _ := json.Marshal(block)
		fmt.Println(string(lastBlockBytes))
		fmt.Println(string(blockBytes))
		fmt.Print("\n-----------\n")

		lastBlockHash := Hash(lastBlock)
		if block.PreviousHash != lastBlockHash {
			return false
		}

		if !ValidProof(lastBlock.Proof, block.Proof, lastBlockHash) {
			return false
		}

		lastBlock = block
		currentIndex++
	}

	return true
}

/*
This is our consensus algorithm, it resolves conflicts by replacing our chain with the longest one in the network.

 return: true if our chain was replaced, false if not
*/
func (b *Blockchain) ResolveConflicts() bool {
	neighbors := b.Nodes
	var newChain []*Block

	maxLength := len(b.Chain)

	for node, _ := range neighbors {
		response, err := http.Get(fmt.Sprintf("http://%s/chain", node))

		if err == nil && response.StatusCode == 200 {
			defer response.Body.Close()
			var result struct {
				Chain  []*Block `json.chain`
				Length int      `json:length`
			}
			json.NewDecoder(response.Body).Decode(&result)
			length := result.Length
			chain := result.Chain

			if length > maxLength && ValidChain(chain) {
				maxLength = length
				newChain = chain
			}
		}
	}

	if newChain != nil {
		b.Chain = newChain
		return true
	}

	return false
}

/*
Create a new Block in the Blockchain

 proof: The proof given by the Proof of Work algorithm
 previousHash: Hash of previous Block
 return: New Block
*/
func (b *Blockchain) NewBlock(proof int64, previousHash string) *Block {
	block := &Block{
		Index:        int32(len(b.Chain) + 1),
		Timestamp:    float64(time.Now().UnixNano()) * 1.0e-9,
		Transactions: b.currentTransactions,
		Proof:        proof,
		PreviousHash: previousHash,
	}

	b.currentTransactions = make([]*Transaction, 0)

	b.Chain = append(b.Chain, block)
	return block
}

/*
Creates a new transaction to go into the next mined Block

 sender: Address of the Sender
 recipient: Address of the Recipient
 amount: Amount
 return: The index of the Block that will hold this transaction
*/
func (b *Blockchain) NewTransaction(sender string, recipient string, amount int64) int32 {
	b.currentTransactions = append(
		b.currentTransactions,
		&Transaction{Sender: sender, Recipient: recipient, Amount: amount})

	return b.LastBlock().Index + 1
}

func (b *Blockchain) LastBlock() *Block {
	return b.Chain[len(b.Chain)-1]
}

/*
Creates a SHA-256 hash of a Block

 block: Block
*/
func Hash(block *Block) string {
	blockBytes, _ := json.Marshal(block)
	hashBytes := sha256.Sum256(blockBytes)
	return hex.EncodeToString(hashBytes[:])
}

/*
Simple Proof of Work Algorithm:

 - Find a number p' such that hash(pp') contains leading 4 zeroes
 - Where p is the previous proof, and p' is the new proof

 lastBlock: last Block
 return:
*/
func ProofOfWork(lastBlock *Block) int64 {
	lastProof := lastBlock.Proof
	lastHash := Hash(lastBlock)

	var proof int64 = 0
	for !ValidProof(lastProof, proof, lastHash) {
		proof++
	}

	return proof
}

/*
Validates the Proof

 lastProof: Previous Proof
 proof: Current Proof
 lastHash: The hash of the Previous Block
 return: true if correct, false if not.
*/
func ValidProof(lastProof int64, proof int64, lastHash string) bool {
	guess := fmt.Sprintf("%d%d%s", lastProof, proof, lastHash)
	guessHashBytes := sha256.Sum256([]byte(guess))
	guessHash := hex.EncodeToString(guessHashBytes[:])
	return strings.HasPrefix(guessHash, "0000")
}
